ratex/csrc/compiler/raf_node_lowering.cpp

```diff
@@ -994,6 +998,117 @@ Var RAFNodeLowering::LowerMean(const ir::ops::Mean* node) {
   return BuildMean({x}, node);
 }
 
+Var BuildMatMul(const std::vector<Var>& ops, const ir::ops::MatMul* node) {
+  LTC_CHECK_EQ(node->operands().size(), 2) << "Unexpected number of operands";
+  Var x = ops[0];
+  Var y = ops[1];
+  Var mm;
+  std::vector<int64_t> a_shape = node->a_shape();
+  std::vector<int64_t> b_shape = node->b_shape();
+  int64_t a_size = a_shape.size();
+  int64_t b_size = b_shape.size();
+
+  if (a_size > 2 && b_size > 2) {
+    if (a_size != 3) {
+      x = BindSymbol(raf::ir::Call(
+          Op::Get("raf.op.reshape"),
+          {x, MakeConstant(TupleInt({-1, a_shape[a_size - 2], a_shape[a_size - 1]}))}));
+    }
+
+    auto trans_axes = lazy_tensors::util::Iota<int64_t>(b_size);
+    std::swap(trans_axes[b_size - 2], trans_axes[b_size - 1]);
+    y = BindSymbol(
+        raf::ir::Call(Op::Get("raf.op.transpose"), {y, MakeConstant(TupleInt(trans_axes))}));
+
+    if (b_size != 3) {
+      y = BindSymbol(raf::ir::Call(
+          Op::Get("raf.op.reshape"),
+          {y, MakeConstant(TupleInt({-1, b_shape[b_size - 1], b_shape[b_size - 2]}))}));
+    }
+
+    mm = BindSymbol(raf::ir::Call(Op::Get("raf.op.batch_matmul_nt"), {x, y}));
+
+    if (a_size != 3) {
+      a_shape.pop_back();
+      a_shape.push_back(b_shape[b_size - 1]);
+      return BindSymbol(
+          raf::ir::Call(Op::Get("raf.op.reshape"), {mm, MakeConstant(TupleInt(a_shape))}));
+    }
+
+    return mm;
+
+  }
+
+  else if (a_size > 2) {
+    x = BindSymbol(raf::ir::Call(Op::Get("raf.op.reshape"),
+                                 {x, MakeConstant(TupleInt({-1, a_shape[a_size - 1]}))}));
+  }
+
+  else if (a_size == 1) {
+    x = BindSymbol(raf::ir::Call(Op::Get("raf.op.expand_dims"), {x, MakeConstant(Int(0))}));
+  }
+
+  if (b_size > 2) {
+    auto trans_axes = lazy_tensors::util::Iota<int64_t>(b_size);
+    std::swap(trans_axes[b_size - 2], trans_axes[b_size - 1]);
+    y = BindSymbol(
+        raf::ir::Call(Op::Get("raf.op.transpose"), {y, MakeConstant(TupleInt(trans_axes))}));
```

Comment:
Ditto. Can we get rid of transpose?

Reply:
Actually, this is a part of the reshape and how we lower >3 dim down to 3dim. We swap the last 2 dims then do a reshape. This way, even after doing the batch_matmul, when we bring back the finals shapes the results are correct.

Reply:
Let me try to understand what you meant. Suppose we have `(m, k) x (b, k, n)`, so you will do `(b, k, n) -> (b, n, k) -> (-1, k) = (b * n, k)`, so that you could apply `dense((m, k), (b * n, k))`.

If the above understanding is correct, you can get rid of the transpose by the following:
1. Expand `a` to be `(1, m, k)`.
2. `batch_matmul(a, b)`.

Reply:
Yes you understanding is correct, and this solution will in fact work. However, as we discussed yesterday (to clue in Zach), the reshape proceeding this will not hold. Instead we will need to add a squeeze or a special reshape for each case whereas the current implementation is more general and the reshapes work correctly for most condition.